// @citation-js/plugin-doi/lib/api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parse = parseDoiApi;
exports.parseAsync = parseDoiApiAsync;

var _core = require("@citation-js/core");

var _json = _interopRequireDefault(require("./json.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const apiOptions = {
  checkContentType: true,
  headers: {
    Accept: 'application/vnd.citationstyles.csl+json'
  }
};

async function fetchDoiApiAsync(url) {
  const result = await _core.util.fetchFileAsync(url, apiOptions);
  return result === '[]' ? {} : JSON.parse(result);
}

async function parseDoiApiAsync(data) {
  const doiJsonList = await Promise.all([].concat(data).map(fetchDoiApiAsync));
  return doiJsonList.map(_json.default);
}

function fetchDoiApi(url) {
  const result = _core.util.fetchFile(url, apiOptions);

  return result === '[]' ? {} : JSON.parse(result);
}

function parseDoiApi(data) {
  return [].concat(data).map(fetchDoiApi).map(_json.default);
}
